pkg/fqdn/messagehandler/message_handler.go

```diff
@@ -175,114 +191,28 @@ func (b *dnsRequestHandler) NotifyOnDNSMsg(
 
 	qname, responseIPs, TTL, CNAMEs, rcode, recordTypes, qTypes, err := dnsproxy.ExtractMsgDetails(msg)
 	if err != nil {
-		b.logger.Error("cannot extract DNS message details",
+		h.logger.Error("cannot extract DNS message details",
 			logfields.Error, err,
 			logfields.DNSName, qname,
 		)
 		return fmt.Errorf("failed to extract DNS message details: %w", err)
 	}
 
 	if msg.Response && msg.Rcode == dns.RcodeSuccess && len(responseIPs) > 0 {
-		stat.PolicyGenerationTime.Start()
-
-		// Create a critical section especially for when multiple DNS requests
-		// are in-flight for the same name (i.e. cilium.io).
-		//
-		// In the absence of such a critical section, consider the following
-		// race condition:
-		//
-		//              G1                                    G2
-		//
-		// T0 --> NotifyOnDNSMsg()               NotifyOnDNSMsg()            <-- T0
-		//
-		// T1 --> UpdateGenerateDNS()            UpdateGenerateDNS()         <-- T1
-		//
-		// T2 ----> mutex.Lock()                 +--------------------------+
-		//                                       |No selectors need updating|
-		// T3 ----> wg := UpdatePolicyMaps()     +--------------------------+
-		//
-		// T4 ----> mutex.Unlock()               mutex.Lock() / mutex.Unlock() <-- T4
-		//
-		// T5 ----> wg.Wait()                    DNS released back to pod    <-- T5
-		//                                                    |
-		// T6 --> DNS released back to pod                    |
-		//              |                                     |
-		//              |                                     |
-		//              v                                     v
-		//       Traffic flows fine                   Leads to policy drop until T6
-		//
-		// Note how G2 releases the DNS msg back to the pod at T5 because
-		// UpdateGenerateDNS() was a no-op. It's a no-op because G1 had executed
-		// UpdateGenerateDNS() first at T1 and performed the necessary policy
-		// updates for the response IPs. Due to G1 performing all the work
-		// first, G2 executes T4 also as a no-op and releases the msg back to the
-		// pod at T5 before G1 would at T6.
-		//
-		// We do not do a `defer unlock()` here, as we should release the lock before
-		// doing final bookkeeping.
-		mutexAcquireStart := time.Now()
-		b.nameManager.LockName(qname)
-
-		if d := time.Since(mutexAcquireStart); d >= option.Config.DNSProxyLockTimeout {
-			b.logger.Warn(fmt.Sprintf("Name lock acquisition time took longer than expected. Potentially too many parallel DNS requests being processed, consider adjusting --%s and/or --%s", option.DNSProxyLockCount, option.DNSProxyLockTimeout),
+		if err := h.UpdateOnDNSMsg(lookupTime, ep, qname, responseIPs, int(TTL), stat); err != nil {
+			h.logger.Error("failed to update DNS message",
```

Comment:
"failed to update DNS message" implies we were updating the contents of a DNS message, but I don't think this is what is happening here. There's also no non-nil error returns from `UpdateOnDNSMsg`. Maybe best to change it to not return an error?

Reply:
Updated